HIVE-27714: Iceberg: metadata location overrides can cause data breach - handling default locations.

ayushtkn opened the pull request:
### What changes were proposed in this pull request?

Add ability to mask default location

### Why are the changes needed?

Better usability  

### Does this PR introduce _any_ user-facing change?

Yes, if configured the auth uri will have default location masked

### Is the change a dependency upgrade?

No

### How was this patch tested?

UT

aturoczy commented on common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:
```diff
@@ -2224,6 +2224,10 @@ public static enum ConfVars {
             " defult DirectExecutorService"),
+
+    HIVE_ICEBERG_MASK_DEFAULT_LOCATION("hive.iceberg.mask.default.location", false,
+        "If set to true the URI for auth will have the default location masked with DEFAULT_TABLE_LOCATION"),
```
If **this is** set to true

ayushtkn replied:
I will change, in the next iter :-)

aturoczy commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -212,6 +216,8 @@ public class HiveIcebergStorageHandler implements HiveStoragePredicateHandler, H
   public static final String MERGE_ON_READ = "merge-on-read";
   public static final String STATS = "/stats/snap-";
 
+  public static final String TABLE_DEFAULT_LOCATION = "TABLE_DEFAULT_LOCATION";
```
What is the TABLE_DEFAULT_LOCATION here? It would not be a default path? Would not be /user/hive/warehouse?

ayushtkn replied:
``TABLE_DEFAULT_LOCATION`` is the constant that we will replace the default location with, say /user/hive/warehouse/database/table1 is the default location, so in the metadata.json path, this will be replaced by ``TABLE_DEFAULT_LOCATION``

so it will look like TABLE_DEFAULT_LOCATION/metadata/abc-metadata.json

That is what Janos use case is: In ranger he can configure policies to allow  paths within ``TABLE_DEFAULT_LOCATION``

aturoczy commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -1038,6 +1050,28 @@ static String encodeString(String rawString) {
 
+  static String getLocation(String location, String defaultLocation, Configuration conf) {
+    if (location == null || defaultLocation == null) {
+      return location;
```
If location is null or defaultLocation is null would not be better to throw an exception? Or it's value checked somewhere else?

ayushtkn replied:
We are typically attempting to mask the location here, if the location is null, we can't mask the default location or there is nothing to mask in the location, if the Client considers it invalid it will throw during auth

aturoczy commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -1038,6 +1050,28 @@ static String encodeString(String rawString) {
+
+  static String getQualifiedPath(String location, Configuration conf) throws IOException {
+    Path path = new Path(location);
+    FileSystem fs = path.getFileSystem(conf);
```
Maybe it is not true for java. But this type of resource should be closable (disposable) in .net. Aka it should be close the resource. I guess there could be some mechanism in Java that could close automatically, but even if it has (don't know) but the best practice is always to close a resource that is not used anymore. Especially those that is coming from the OS

ayushtkn replied:
FileSystem is cached, it isn't created always, it is created once & next time when you try to fetch it, it returns the same instance not a new one. In the end, all the FS instances are closed for the UGI which were created by Hive during the query execution, so there isn't no explicit need for closing the FS, it will get closed in the end.
If we close it in the middle, similar issues will arise like in: [HIVE-27447](https://issues.apache.org/jira/browse/HIVE-27447)

aturoczy replied:
Ack

aturoczy commented on ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:
```diff
@@ -14215,17 +14216,18 @@ ASTNode analyzeCreateTable(
         }
-      } catch (HiveException e) {
-        throw new RuntimeException(e);
+      } catch (HiveException | MetaException e) {
```
Rarely see the pipe in exception :) 👍

aturoczy reviewed (changes requested):
Minor changes + some small questions

ayushtkn commented:
I have addressed the comments. There was a test failure, it was basically the path was specified as unqualified & the output was qualified.
So, in case we don't do the masking, we should make sure that if the path specified is Qualified, the same is returned, if it isn't qualified the same form should be returned.

Even added a check the paths should be in same FS. 

Bit of if-else now, will await for further reviews :-)

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -1011,25 +1018,31 @@ public URI getURIForAuth(org.apache.hadoop.hive.metastore.api.Table hmsTable) th
         SessionStateUtil.getProperty(conf, BaseMetastoreTableOperations.METADATA_LOCATION_PROP);
     if (metadataLocation.isPresent()) {
-      authURI.append(encodeString(metadataLocation.get()));
+      authURI.append(encodeString(getLocation(metadataLocation.get(), getDefaultCreateLocation(), conf)));
```
could you please refactor `getLocation(....)` to `getPathForAuth(String path)`
````
return encodeString(location)
````

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -1011,25 +1018,31 @@ public URI getURIForAuth(org.apache.hadoop.hive.metastore.api.Table hmsTable) th
     return new URI(authURI.toString());
   }
 
+  private String getDefaultCreateLocation() {
```
no need for this method, just use it inside `getPathForAuth`

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -1038,6 +1051,44 @@ static String encodeString(String rawString) {
+    }
+    boolean maskDefaultLocation = conf.getBoolean(HiveConf.ConfVars.HIVE_ICEBERG_MASK_DEFAULT_LOCATION.varname,
+        HiveConf.ConfVars.HIVE_ICEBERG_MASK_DEFAULT_LOCATION.defaultBoolVal);
+    if (maskDefaultLocation) {
```
````
if (!maskDefaultLocation) {
  return location;
}
````

deniskuzZ replied:
why not remove nesting?
````
if (!maskDefaultLocation) {
  return location;
}
````

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -1011,19 +1018,19 @@ public URI getURIForAuth(org.apache.hadoop.hive.metastore.api.Table hmsTable) th
         SessionStateUtil.getProperty(conf, BaseMetastoreTableOperations.METADATA_LOCATION_PROP);
     if (metadataLocation.isPresent()) {
-      authURI.append(encodeString(metadataLocation.get()));
+      authURI.append(getPathForAuth(metadataLocation.get(), null, conf));
```
you don't need to pass `conf`, it's an instance variable. remove static and make it private. Also, please create a default method with 1 arg.

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -1640,7 +1681,17 @@ public boolean shouldOverwrite(org.apache.hadoop.hive.ql.metadata.Table mTable,
+    Optional<String> defaultTableLocation =
+        SessionStateUtil.getProperty(conf, hive_metastoreConstants.DEFAULT_TABLE_LOCATION);
+    if (!defaultTableLocation.isPresent()) {
+      Warehouse wh = new Warehouse(conf);
```
why do you need all of this? in SA#analyzeCreateTable there is already available tblLocation.
````
storageHandler.addResourcesForCreateTable(tblProps, conf, tblLocation)
````

deniskuzZ replied:
you can just extract in SA method `getDefaultTablePath`
````
        try {
          Warehouse wh = new Warehouse(conf);
          tblLocation = wh.getDefaultTablePath(db.getDatabase(qualifiedTabName.getDb()), qualifiedTabName.getTable(),
                  isExt).toUri().getPath();
        } catch (MetaException | HiveException e) {
          throw new SemanticException(e);
        } 
````

ayushtkn replied:
Done, refactored the method & called from validateAndAddDefaultProperties

nrg4878 commented:
@ayushtkn  the hive_metastoreConstants is a generated class. We cannot make change to it directly. Any changes will have to be made to hive_metastore.thrift file.

So this breaks the build if you do a clean build

ayushtkn commented:
Thanx @nrg4878 for pointing it out, I have quickly fixed it in addendum PR at https://github.com/apache/hive/pull/4921

If you can help check if it is working fine for you. Not sure why the build on this PR didn't fail nor did the subsequent PRs, maybe you were using Pthrift and the build doesn't use that, I can followup on that as well & make sure it doesn't happen next time.

In case the addendum PR doesn't make sense to you or has issues or doesn't fix the build. I can revert it or if it blocks you feel free to revert this commit & I can rework
